src/device/lib.rs

```diff
@@ -158,39 +161,59 @@ impl<T: Copy, I> Handle<T, I> {
 }
 
 /// Type-safe buffer handle
-#[derive(Copy, Debug, PartialEq, Clone)]
-pub struct BufferHandle<T> {
-    raw: RawBufferHandle,
+pub struct BufferHandle<R: Resources, T> {
+    raw: RawBufferHandle<R>,
+}
+
+impl<R: Resources, T> Copy for BufferHandle<R, T> {}
+
+impl<R: Resources, T> Clone for BufferHandle<R, T> {
+    fn clone(&self) -> BufferHandle<R, T> {
+        BufferHandle { raw: self.raw }
+    }
+}
+
+impl<R: Resources, T> PartialEq for BufferHandle<R, T> {
+    fn eq(&self, other: &BufferHandle<R, T>) -> bool {
+        self.raw == other.raw
+    }
+}
+
+impl<R: Resources, T> fmt::Debug for BufferHandle<R, T> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "BufferHandle {{ raw: {:?} }}", self.raw)
+    }
 }
 
-impl<T> BufferHandle<T> {
+impl<R: Resources, T> BufferHandle<R, T> {
     /// Create a type-safe BufferHandle from a RawBufferHandle
-    pub fn from_raw(handle: RawBufferHandle) -> BufferHandle<T> {
+    pub fn from_raw(handle: RawBufferHandle<R>) -> BufferHandle<R, T> {
         BufferHandle {
             raw: handle,
         }
     }
 
     /// Cast the type this BufferHandle references
-    pub fn cast<U>(self) -> BufferHandle<U> {
+    pub fn cast<U>(self) -> BufferHandle<R, U> {
         BufferHandle::from_raw(self.raw)
     }
 
-    /// Get the underlying GL name for this BufferHandle
-    pub fn get_name(&self) -> back::Buffer {
+    /// Get the underlying name for this BufferHandle
+    pub fn get_name(&self) -> <R as Resources>::Buffer {
```

Comment:
Are you sure casting here is required? I mean, it might as well be, but Rust should be smart enough (eventually) to see what `R::` has.